numpy/fft/_pocketfft_umath.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * This file is part of pocketfft.
+ * Licensed under a 3-clause BSD style license - see LICENSE.md
+ */
+
+/*
+ *  Main implementation file.
+ *
+ *  Copyright (C) 2004-2018 Max-Planck-Society
+ *  \author Martin Reinecke
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#define PY_SSIZE_T_CLEAN
+#include <assert.h>
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/ufuncobject.h"
+
+#include "npy_config.h"
+
+#include "pocketfft/pocketfft.h"
+
+/*
+ * Copy all nin elements of input to the first nin of the output,
+ * and any set any remaining nout-nin output elements to 0
+ * (if nout < nin, copy only nout).
+ */
+
+static inline void
+copy_data(char* in, npy_intp step_in, npy_intp nin,
+          char* out, npy_intp step_out, npy_intp nout, npy_intp elsize)
+{
+    npy_intp ncopy = nin <= nout? nin : nout;
+    if (step_in == elsize && step_out == elsize) {
+        memcpy(out, in, ncopy*elsize);
+    } else {
+        char *ip = in, *op = out;
+        for (npy_intp i = 0; i < ncopy; i++, ip += step_in, op += step_out) {
+            memcpy(op, ip, elsize);
+        }
+    }
+    if (nin < nout) {
+        char *op = out + nin*elsize;
+        if (step_out == elsize) {
+            memset(op, 0, (nout-nin)*elsize);
+        }
+        else {
+            for (npy_intp i = nin; i < nout; i++, op += step_out) {
+                memset(op, 0, elsize);
+            }
+        }
+    }
+}
+
+
+/*
+ * Loops calling the pocketfft code.
+ *
+ * Unfortunately, the gufunc machinery does not (yet?) allow forcing contiguous
+ * inner loop data, so we create a contiguous output buffer if needed
+ * (input gets copied to output before processing, so can be non-contiguous).
+ */
+static void
+fft_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
```

Comment:
Heh, I actually don't remember that the outer strides come all before the inner ones (although it does make sense).  Tests should fail left and right if not, though.

(Can't say I found the name `isi` with the first `i` being inner immediately obvious. OTOH, something long with `core_stride_in` or so is maybe long below, so doesn't matter.)

Reply:
Yes, this is right, and, yes, the names are awful. I think they were used in the `_test_umath` routine that I started with... (a simple example ufunc module is so useful...). Changed them to `step_in` and `step_out`, now, to match what I have for `copy_data`.